File: code_samples/section12/example_7_bubble_sort/bubble_sort.c
#include <stdio.h>   // printf, fopen, fscanf
#include <stdlib.h>  // malloc, free
#include <stdbool.h> // bool, true, false
#include <string.h>  // (not used heavily here, but included)

// ------------------------------------------------------------
// Bubble Sort with Step Counting
// ------------------------------------------------------------
/*
    Bubble sort repeatedly passes through the array, swapping adjacent
    out-of-order elements. After each full pass, the largest remaining
    element "bubbles" to the end of the unsorted portion.

    This implementation includes an optimization:
      - If a full pass makes NO swaps, the array is already sorted and
        we can stop early.

    Step counting:
      - *comparisons counts how many adjacent comparisons we perform:
            arr[i-1] > arr[i]
      - *swaps counts how many swaps we actually execute

    Parameters:
      - arr: array to sort (modified in place)
      - n: number of elements
      - comparisons: output counter for comparisons
      - swaps: output counter for swaps
*/
void bubble_sort(int* arr, int n, long long* comparisons, long long* swaps) {
    bool swapped = true;  // controls whether we need another pass

    // Initialize counters
    *comparisons = 0;
    *swaps = 0;

    // Continue making passes while we keep swapping something
    while (swapped) {
        swapped = false;  // assume no swaps this pass unless we perform one

        // One pass compares adjacent pairs from index 1..n-1
        for (int i = 1; i < n; i++) {
            (*comparisons)++;  // Count comparison of one adjacent pair

            // If elements are out of order, swap them
            if (arr[i - 1] > arr[i]) {
                int tmp = arr[i - 1];
                arr[i - 1] = arr[i];
                arr[i] = tmp;

                (*swaps)++;    // Count swap
                swapped = true; // Signal that we need another pass
            }
        }

        // After each pass, the last element of the current range is in its
        // final sorted position, so we can reduce the range by 1.
        n--; // last element is in correct position
    }
}

// ------------------------------------------------------------
// Helper: Load integers from a file
// ------------------------------------------------------------
/*
    Loads a whitespace-delimited list of integers from a file.

    Strategy:
      1) Count how many integers are in the file.
      2) Allocate an array of exactly that size.
      3) Rewind and read the integers into the array.

    Parameters:
      - filename: path to the file to read
      - outArr: output pointer that receives the allocated array

    Returns:
      - number of integers read on success
      - -1 on failure

    Caller responsibility:
      - Must free(*outArr) when done.
*/
int load_file(const char* filename, int** outArr) {
    FILE* f = fopen(filename, "r");
    if (!f) {
        printf("Error reading: %s\n", filename);
        return -1;
    }

    // -----------------------------
    // First pass: count numbers
    // -----------------------------
    int count = 0, tmp;
    while (fscanf(f, "%d", &tmp) == 1) {
        count++;
    }

    // Return to start of file for second pass
    rewind(f);

    // Allocate enough space for all integers
    int* arr = malloc(count * sizeof(int));

    // Second pass: read values into the allocated array
    for (int i = 0; i < count; i++) {
        fscanf(f, "%d", &arr[i]);
    }

    fclose(f);

    // Provide the allocated array to the caller
    *outArr = arr;
    return count;
}

// ------------------------------------------------------------
// Compare two arrays for equality
// ------------------------------------------------------------
/*
    Compares two arrays element-by-element and counts mismatches.

    Also prints up to the first 10 mismatches for debugging so output
    doesn't become overwhelming.

    Parameters:
      - a: first array
      - b: second array (reference / expected)
      - n: number of elements to compare

    Returns:
      - number of mismatched positions found
*/
int compare_arrays(int* a, int* b, int n) {
    int mismatches = 0;

    for (int i = 0; i < n; i++) {
        if (a[i] != b[i]) {
            // Print only a limited number of mismatches
            if (mismatches < 10)
                printf("Mismatch at %d: got %d, expected %d\n", i, a[i], b[i]);
            mismatches++;
        }
    }

    return mismatches;
}

// ------------------------------------------------------------
// Main test driver
// ------------------------------------------------------------
/*
    Test workflow:
      1) Load unordered.txt (unsorted input data)
      2) Load ordered.txt   (reference sorted output)
      3) Run bubble sort on unordered data, counting comparisons & swaps
      4) Compare the sorted result against the reference
      5) Print statistics and success/failure summary

    Notes on paths:
      - Uses relative paths like "../data/unordered.txt"
      - These paths depend on where the program is run from.
*/
int main(void) {
    const char* unordered_path = "../data/unordered.txt";
    const char* ordered_path   = "../data/ordered.txt";

    int *unordered = NULL, *ordered = NULL;

    // -----------------------------
    // Load unordered input
    // -----------------------------
    printf("Attempting to read: %s\n", unordered_path);
    int n1 = load_file(unordered_path, &unordered);
    if (n1 <= 0) {
        printf("Missing unordered file - aborting.\n");
        return 1;
    }

    // -----------------------------
    // Load ordered reference output
    // -----------------------------
    printf("Attempting to read: %s\n", ordered_path);
    int n2 = load_file(ordered_path, &ordered);
    if (n2 <= 0) {
        printf("Missing ordered file - aborting.\n");
        free(unordered);
        return 1;
    }

    // Sanity check: both files should have the same number of values
    if (n1 != n2) {
        printf("File length mismatch - aborting.\n");
        free(unordered);
        free(ordered);
        return 1;
    }

    printf("\n--- Bubble Sort Step Count ---\n");

    // Counters for algorithm work performed
    long long comparisons = 0, swaps = 0;

    // Sort the unordered data in place
    bubble_sort(unordered, n1, &comparisons, &swaps);

    // Report step counts
    printf("Comparisons: %lld\n", comparisons);
    printf("Swaps:       %lld\n", swaps);

    // -----------------------------
    // Verify correctness
    // -----------------------------
    printf("\nComparing result to ordered.txt...\n");

    int mismatches = compare_arrays(unordered, ordered, n1);

    if (mismatches == 0)
        printf("SUCCESS: Output matches ordered.txt\n");
    else
        printf("FAIL: %d mismatches found.\n", mismatches);

    // Clean up allocated memory
    free(unordered);
    free(ordered);

    return 0;
}

File: code_samples/section12/example_7_bubble_sort/bubble_sort.cs
#nullable enable

// =============================================================
// Bubble Sort (baseline / no step counting)
// =============================================================
//
// This is a plain bubble sort implementation that:
//   - Repeatedly sweeps through the array,
//   - Swaps adjacent out-of-order elements,
//   - Stops early when a full pass completes with no swaps.
//
// Notes:
//   - The variable `n` shrinks each pass because the largest element in
//     the current unsorted range "bubbles" to the end of that range.
//   - This function sorts the array IN PLACE.
//
static void BubbleSort(int[] arr) {
    int n = arr.Length;
    bool swapped = true;

    // Keep running passes until a pass makes no swaps
    while (swapped) {
        swapped = false;

        // Compare adjacent pairs within the current unsorted range [0..n-1]
        for (int i = 1; i < n; i++) {
            if (arr[i - 1] > arr[i]) {
                // Tuple swap syntax (compact swap of two array elements)
                (arr[i], arr[i - 1]) = (arr[i - 1], arr[i]);
                swapped = true;
            }
        }

        // After each full pass, the element at index (n-1) is finalized
        n--;
    }
}

// -------------------------------------------------------------
// Bubble Sort with step counting
// -------------------------------------------------------------
//
// This version adds a StepCounter object to track work performed.
//
// Step definitions used here:
//   - Comparisons: number of adjacent comparisons performed
//       arr[i - 1] > arr[i]
//   - Swaps: number of actual swaps executed
//
// Important behavior notes:
//   - Sorts IN PLACE.
//   - Uses the same early-exit optimization:
//       if no swaps occur in a pass, the array is already sorted.
//   - Shrinks the effective range by 1 after each pass (n--).
//
void BubbleSortSteps(int[] arr, StepCounter stats)
{
    int n = arr.Length;
    bool swapped = true;

    // Repeat passes until the array is sorted (no swaps in a pass)
    while (swapped)
    {
        swapped = false;

        // One pass through the current unsorted portion
        for (int i = 1; i < n; i++)
        {
            // Count the adjacent comparison performed each iteration
            stats.Comparisons++;

            // Swap if the pair is out of order
            if (arr[i - 1] > arr[i])
            {
                // Manual swap using a temp variable
                int tmp = arr[i - 1];
                arr[i - 1] = arr[i];
                arr[i] = tmp;

                // Count the swap and mark that a swap occurred this pass
                stats.Swaps++;
                swapped = true;
            }
        }

        // Last element in the current range is now in the correct place,
        // so we don't need to include it in subsequent passes.
        n--; // Last element is now in the correct place
    }
}

// -------------------------------------------------------------
// Auto-detect project root (fixes dotnet-script path problems)
// -------------------------------------------------------------
//
// dotnet-script / top-level scripts often run with a working directory
// that isn't the repo root. This helper walks up parent directories
// until it finds a folder that contains "code_samples".
//
// Returns:
//   - Path to detected project root if found
//   - Current directory as a fallback if not found
//
string FindProjectRoot()
{
    string cur = Directory.GetCurrentDirectory();

    // Climb upward until we reach the filesystem root
    while (cur != Directory.GetDirectoryRoot(cur))
    {
        // Heuristic: repo root is assumed to contain "code_samples"
        if (Directory.Exists(Path.Combine(cur, "code_samples")))
        {
            return cur;
        }

        // Move one directory up
        cur = Directory.GetParent(cur)!.FullName;
    }

    // fallback if nothing found
    return Directory.GetCurrentDirectory();
}

// -------------------------------------------------------------
// Load a file from code_samples/section12/data
// -------------------------------------------------------------
//
// Loads a whitespace-delimited integer file from:
//   <project-root>/code_samples/section12/data/<filename>
//
// Returns:
//   - int[] on success
//   - null on failure (file missing, unreadable, parse error, etc.)
//
// Also prints the exact path being attempted to help debug path issues.
//
int[]? LoadFile(string filename)
{
    try
    {
        // Determine the repo/project root robustly
        string root = FindProjectRoot();

        // Build the full path to the requested data file
        string fullPath = Path.Combine(
            root,
            "code_samples",
            "section12",
            "data",
            filename
        );

        Console.WriteLine($"Attempting to read: {fullPath}");

        // Read entire file and parse ints split on whitespace
        string text = File.ReadAllText(fullPath);
        return [.. text
            .Split((char[])null!, StringSplitOptions.RemoveEmptyEntries)
            .Select(int.Parse)];
    }
    catch (Exception ex)
    {
        Console.WriteLine($"Error reading: {filename}");
        Console.WriteLine(ex.Message);
        return null;
    }
}

// -------------------------------------------------------------
// Compare arrays for mismatches
// -------------------------------------------------------------
//
// Compares two arrays element-by-element up to the smaller length
// (Math.Min is used as a defensive measure, even though the caller
// typically checks for equal lengths first).
//
// Prints up to the first 10 mismatches for readability.
//
// Returns:
//   - number of mismatched positions found
//
int CompareArrays(int[] a, int[] b)
{
    int mismatches = 0;
    int n = Math.Min(a.Length, b.Length);

    for (int i = 0; i < n; i++)
    {
        if (a[i] != b[i])
        {
            // Print only the first few mismatches to avoid flooding output
            if (mismatches < 10)
                Console.WriteLine($"Mismatch @ {i}: got {a[i]}, expected {b[i]}");

            mismatches++;
        }
    }

    return mismatches;
}

// -------------------------------------------------------------
// Top-level program (NO Main function)
// -------------------------------------------------------------
//
// Workflow:
//   1) Load unordered.txt (input to sort)
//   2) Load ordered.txt   (expected sorted output reference)
//   3) Run bubble sort with step counting on unordered data
//   4) Compare result to expected output and report success/failure
//
var unordered = LoadFile("unordered.txt");
var expected  = LoadFile("ordered.txt");

// Validate file loading
if (unordered is null || expected is null)
{
    Console.WriteLine("Missing input file — aborting.");
    return;
}

// Validate lengths match (so element-by-element comparison is meaningful)
if (unordered.Length != expected.Length)
{
    Console.WriteLine("File length mismatch — aborting.");
    Console.WriteLine($"unordered: {unordered.Length}, expected: {expected.Length}");
    return;
}

// Create a counter object to record comparisons and swaps
var stats = new StepCounter();

Console.WriteLine("\n--- Bubble Sort Step Count ---");

// Sort IN PLACE and accumulate step counts
BubbleSortSteps(unordered, stats);

// Print step-count results
Console.WriteLine($"Comparisons: {stats.Comparisons}");
Console.WriteLine($"Swaps:       {stats.Swaps}");

// Verify correctness by comparing to expected sorted output
Console.WriteLine("\nChecking sorted output...");
int mismatchesFound = CompareArrays(unordered, expected);

if (mismatchesFound == 0)
    Console.WriteLine("SUCCESS — output matches expected sorted list!");
else
    Console.WriteLine($"FAIL — mismatches found: {mismatchesFound}");

// -------------------------------------------------------------
// StepCounter record type
// -------------------------------------------------------------
//
// Simple container for step-count statistics.
// Using a record here provides a compact syntax while still allowing
// mutable properties.
//
record StepCounter
{
    public long Comparisons { get; set; }
    public long Swaps { get; set; }
}
